### ISSUE TITLE
Attachment issues

### ISSUE THREAD

**Issue by Karitham** [wneessen/go-mail#110]
### Description

Multiple calls to `WriteTo` lose files (and embeds).

### To Reproduce

Call `WriteTo` multiple times.

for example, `msg.WriteFile` and then `client.Send(msg)`

### Expected behaviour

Attachments are written with every call to `WriteTo`

### Attempted Fixes

I haven't attempted any fix yet, but expect that breaking the `io.Reader` interface to an `io.ReadSeeker` would be the simplest fix to that issue.

Another way to avoid breaking compatibility would be to call the `fileOptions` on every call to write, and set the writer in a closure, like this.

```go
		msg.AttachFile(a.Name, func(f *mail.File) {
			buf := bytes.NewBuffer(a.Content)
			f.Writer = func(w io.Writer) (int64, error) {
				return buf.WriteTo(w)
			}
			f.Name = a.Name
			f.Header.Set(string(mail.HeaderContentType), a.ContentType)
		})
```

If the options are called on `WriteTo`, subsequent calls to `WriteTo` get a new writer, or the user can try to seek if that's better.

**Comment by wneessen** [comment]
Hi @Karitham,

thanks for the report but I can't reproduce the behaviour you are experiencing.

I've quickly used this test code:
```go
package main

import (
	"fmt"
	"os"

	"github.com/wneessen/go-mail"
)

func main() {
	m := mail.NewMsg()
	m.Subject("Subject")
	_ = m.To("wn@neessen.dev")
	_ = m.From("go-mail@neessen.net")
	m.SetBodyString(mail.TypeTextPlain, "Content")
	m.AttachFile("main.go")
	m.EmbedFile("main.go")

	fmt.Println("First try:")
	_, _ = m.WriteTo(os.Stdout)
	fmt.Println("Second try:")
	_, _ = m.WriteTo(os.Stdout)
	fmt.Println("Third try:")
	_ = m.WriteToFile("tempfile.out")

	fmt.Println("Send mail:")
	tu := os.Getenv("TEST_USER")
	tp := os.Getenv("TEST_PASS")
	c, err := mail.NewClient(os.Getenv("TEST_HOST"), mail.WithTLSPolicy(mail.TLSMandatory),
		mail.WithSMTPAuth(mail.SMTPAuthPlain), mail.WithUsername(tu),
		mail.WithPassword(tp), mail.WithPort(587))
	if err != nil {
		fmt.Printf("failed to create new client: %s\n", err)
		os.Exit(1)
	}
	if err := c.DialAndSend(m); err != nil {
		fmt.Printf("failed to dial: %s\n", err)
		os.Exit(1)
	}
}
```

And both the embed and the attached files are present in all 4 outputs (meaning, the two `msg.WriteTo()` to `os.Stdout`, the `msg.WriteToFile()` and the final sending via a mail server to an actual mail account.

Is it possible that you are using an old version of go-mail. I remembers that we had an issue with consecutive writes in a previous version. If you are not using the latest version, can you please retry with v0.3.8, please? In case you are using the latest version, would you be able to provide a proof of concept code snippet that shows the issue in action? Maybe also provide some environment details like OS/Go version, so I can try to reproduce?

**Comment by Karitham** [comment]
```go
func main() {
	m := mail.NewMsg()
	m.AttachReader("hi", bytes.NewBuffer([]byte("oh hi mark!")))

	o1 := bytes.Buffer{}
	m.WriteTo(&o1)

	o2 := bytes.Buffer{}
	m.WriteTo(&o2)

	fmt.Println(len(o1.String()), len(o2.String()))
	fmt.Println("----")
	fmt.Println(o1.String())
	fmt.Println("----")
	fmt.Println(o2.String())
}
```

**Comment by wneessen** [comment]
Thanks, again. This helps. I'll have a deeper look soon.

**Comment by wneessen** [comment]
This has been fixed with PR #111. To not break the current implementation that requires an `io.Reader` the data is read into memory and then buffered in a `io.ReadSeeker`. Additionally, there we now have `Msg.AttachReadSeeker()` and `Msg.EmbedReadSeeker()` which require an `io.ReadSeeker` interface to be provided as argument, which should be more efficient since the reading into memory is skipped and left to the `io.Seeker`.

We are just in the mids of finishing some features, so you can expect a new release containing this fix soon. In the meantime you can work with `main`.
